serialization/cbor.ksy

```diff
@@ -0,0 +1,328 @@
+            if: exponent_valid
+          mantissa:
+            value: "arr.items[1].payload.as<uint>.value"
+            if: mantissa_valid
+          valid:
+            value: exponent_valid and mantissa_valid
+      # Kaitai has neither inheritance no duck typing, so have to repeat code
+      big_float:
+        instances:
+          array_valid:
+            value: _parent.as<tag>.tagged.ib.major_type == type::array
+          arr:
+            value: _parent.as<tag>.tagged.payload.as<array>
+            if: array_valid
+          array_length_valid:
+            value: "array_valid and arr.length.value == 2"
+          bfc:
+            pos: 0
+            type: big_float_common(arr)
+            if: array_valid and array_length_valid
+          scale:
+            pos: 0
+            type: pow2(bfc.exponent)
+            # a dirty and limited workaround for Kaitai not having a power operation. The limitation is: exponent is only allowed to be 64 bits
+          float_approximation:
+            value: "bfc.mantissa*scale.value"
+            if: bfc.valid
+        types:
+          pow2:
+            params:
+              - id: log
+                type: u8
+            instances:
+              is_positive:
+                value: log>=0
+              mod:
+                value: (is_positive?log:-log)
+              positive_pow:
+                value: "1<<mod"
+              value:
+                value: (is_positive?positive_pow:1./positive_pow)
+      decimal:
+        instances:
+          array_valid:
+            value: _parent.as<tag>.tagged.ib.major_type == type::array
+          arr:
+            value: _parent.as<tag>.tagged.payload.as<array>
+            if: array_valid
+          array_length_valid:
+            value: "array_valid and arr.length.value == 2"
+          bfc:
+            pos: 0
+            type: big_float_common(arr)
+            if: array_valid and array_length_valid
+          scale:
+            pos: 0
+            type: pow10(bfc.exponent)
+            # a dirty and broken workaround for Kaitai not having a power operation. Has limitations: exponent is only allowed to be 8 bit
+          float_approximation:
+            value: "bfc.mantissa*scale.value"
+            if: bfc.valid
+        types:
+          pow10:
+            params:
+              - id: log
+                type: u1
+            instances:
+              is_positive:
+                value: log>=0
+              mod:
+                value: (is_positive?log:-log)
+              positive_pow:
+                value: "(((mod & 1).as<b1> ? 5 : 1) * ((mod & 2).as<b1> ? 25 : 1) * ((mod & 4).as<b1> ? 625 : 1) * ((mod & 8).as<b1> ? 390625 : 1) * ((mod & 16).as<b1> ? 0x2386f26fc1 : 1) * ((mod & 32).as<b1> ? 0x4ee2d6d415b85acef81 : 1) * ((mod & 64).as<b1> ? 0x184f03e93ff9f4daa797ed6e38ed64bf6a1f01 : 1) * ((mod & 128).as<b1> ? 0x24ee91f2603a6337f19bccdb0dac404dc08d3cff5ec2374e42f0f1538fd03df99092e953e01 : 1)) << mod"
```

Comment:
Simply put, casting `CalcIntType` to `b1` is evil. The type casting in KS is currently quite dumb: in dynamically typed languages it's effectively noop (just fools the type checking engine in the compiler), in static typed languages it blindly does the explicit casting without checking whether the types are compatible or not. We really need to fix that (https://github.com/kaitai-io/kaitai_struct/issues/696) 

And I just checked that `(boolean) 4` fails at least in Java, because `incompatible types: int cannot be converted to boolean`. So please use `(mod & 1) != 0`, as it is the only reliable way.

Another thing is that we don't have integers with infinite precision (`BigInteger`) in KS (that's also the reason why the power operation is not yet implemented: https://github.com/kaitai-io/kaitai_struct/issues/216), so the number can easily overflow in most languages (except Python, of course). Even if we had power operation that takes and returns `double`s, it apparently isn't enough in this case. _(Hmm, I can't find an dedicated issue for bignums, it's shortly mentioned in https://github.com/kaitai-io/kaitai_struct/issues/104#issuecomment-358261907 and support for 64-bit integers in JS is related: https://github.com/kaitai-io/kaitai_struct/issues/183 but not quite. No one seems to consider it so important.)_

Actually, I'm surprised that the compiler written in Scala is capable of dealing with such big integers like `0x4ee2d6d415b85acef81`, that's nice.

Reply:
Thanks.